File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    private int subsequence(String text1,String text2,int n,int m,int dp[][]){
        if(n==0 || m ==0){
            return 0;
        }
        if(dp[n-1][m-1] !=-1){
            return dp[n-1][m-1];
        }
        if(text1.charAt(n-1) == text2.charAt(m-1)){ 
            System.out.println(text1.charAt(n-1));
            return dp[n-1][m-1] = 1 + subsequence(text1,text2,n-1,m-1,dp);
        }
        return dp[n-1][m-1]= 0 + Math.max(subsequence(text1,text2,n-1,m,dp),subsequence(text1,text2,n,m-1,dp));
    }
    public int longestCommonSubsequence(String text1, String text2) {
        int n = text1.length();
        int m = text2.length();
        int dp[][] = new int[n][m];
        for(int row[]: dp){
            Arrays.fill(row,-1);
        }
        return subsequence(text1,text2,n ,m,dp);
    }
}

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
private:
    int f(int ind1,int ind2,string &s1,string &s2,vector<vector<int>> &dp) {
        if(ind1 < 0 || ind2 < 0) return 0;
        if(dp[ind1][ind2] != -1) return dp[ind1][ind2];
        if(s1[ind1] == s2[ind2]) {
            return dp[ind1][ind2] = 1 + f(ind1-1,ind2-1,s1,s2,dp);
        }
        return dp[ind1][ind2] = max(f(ind1-1,ind2,s1,s2,dp) , f(ind1,ind2-1,s1,s2,dp));
    }
public:
    int longestCommonSubsequence(string s1, string s2) {
        int n = s1.size();
        int m = s2.size();
        vector<vector<int>> dp(n,vector<int>(m,-1));
        return f(n-1,m-1,s1,s2,dp);
    }
};